internal/consensus/pbts_test.go

```diff
@@ -531,3 +534,163 @@ func pbtsFromHeightParams(height int64) types.FeatureParams {
+			block, blockParts = rs.ProposalBlock, rs.ProposalBlockParts
+			blockID = rs.Proposal.BlockID
+		} else {
+			var ts time.Time
+			if height >= pbtsSetHeight && height < pbtsEnableHeight {
+				// Use PBTS logic while PBTS is not yet activated
+				ts = time.Now()
+				rejectProposal = true
+			} else if height >= pbtsEnableHeight {
+				// Shift timestamp to the future => not timely
+				ts = time.Now().Add(time.Second)
+				rejectProposal = true
+			}
+			block, blockParts, blockID = createProposalBlock(t, cs, ts)
+			proposal := types.NewProposal(height, round, -1, blockID, block.Header.Time)
+			// BFT Time should not care about Proposal's timestamps
+			if height < pbtsSetHeight {
+				proposal.Timestamp = time.Now()
+			}
+			signProposal(t, proposal, chainID, vss[proposer])
+			cs.SetProposalAndBlock(proposal, block, blockParts, "p")
+			ensureNewProposal(proposalCh, height, round)
+		}
+
+		delta := time.Now().Sub(block.Time)
+		t.Log("BLOCK", height, round, "PROPOSER", proposer, "PBTS", pbtsEnabled,
+			"TIMESTAMP", block.Time, delta, "ACCEPTED", !rejectProposal)
+
+		// Accept proposal and decide, or reject proposal and move to next round
+		myVote := blockID.Hash
+		lockedRound := round
+		if rejectProposal {
+			myVote = nil
+			lockedRound = int32(-1)
+		} else { // We are deciding, enable FinalizeBlock mock
+			res := &abci.FinalizeBlockResponse{}
+			// Enable PBTS from pbtsEnableHeight via consensus params
+			if height == pbtsSetHeight {
+				params := types.DefaultConsensusParams()
+				params.Feature.VoteExtensionsEnableHeight = 1
+				params.Feature.PbtsEnableHeight = pbtsEnableHeight
+				paramsProto := params.ToProto()
+				res.ConsensusParamUpdates = &paramsProto
+			}
+			app.On("FinalizeBlock", mock.Anything, mock.Anything).
+				Return(res, nil).Once()
+		}
+
+		// Prevote step
+		ensurePrevote(voteCh, height, round)
+		validatePrevote(t, cs, round, vss[0], myVote)
+		for _, vs := range vss[2:] {
+			signAddVotes(cs, types.PrevoteType, chainID, blockID, false, vs)
+			ensurePrevote(voteCh, height, round)
+		}
+
+		// Precommit step
+		ensurePrecommit(voteCh, height, round)
+		validatePrecommit(t, cs, round, lockedRound, vss[0], myVote, myVote)
+		for _, vs := range vss[2:] {
+			ts := time.Now()
+			// Shift the next block timestamp while running BFT Time
+			if height >= pbtsSetHeight-1 && height < pbtsEnableHeight {
+				ts = ts.Add(time.Second)
+			}
+			vote := signVoteWithTimestamp(vs, types.PrecommitType, chainID, blockID, true, ts)
+			cs.peerMsgQueue <- msgInfo{Msg: &VoteMessage{vote}}
+			ensurePrecommit(voteCh, height, round)
+		}
+
+		if myVote != nil {
+			height, round = height+1, 0
+			incrementHeight(vss[1:]...)
+		} else {
+			height, round = height, round+1
+			incrementRound(vss[1:]...)
+		}
+	}
+	// Last call to FinalizeBlock
+	ensureNewRound(newRoundCh, height, round)
+}
+
+func createProposalBlock(t *testing.T, cs *State, timestamp time.Time) (*types.Block, *types.PartSet, types.BlockID) {
```

Comment:
Have you looked around to see if we have similar utility functions elsewhere in the consensus UTs that we could reuse? Same for `signProposal` below.